lib/Conversion/TritonGPUToLLVM/ElementwiseOpToLLVM.cpp

```diff
@@ -512,8 +512,104 @@ class ElementwiseOpConversionBase
   using OpAdaptor = typename SourceOp::Adaptor;
 
   explicit ElementwiseOpConversionBase(
-      TritonGPUToLLVMTypeConverter &typeConverter, PatternBenefit benefit = 1)
-      : ConvertTritonGPUOpToLLVMPattern<SourceOp>(typeConverter, benefit) {}
+      TritonGPUToLLVMTypeConverter &typeConverter,
+      ModuleAxisInfoAnalysis &axisAnalysisPass, PatternBenefit benefit = 1)
+      : ConvertTritonGPUOpToLLVMPattern<SourceOp>(typeConverter, benefit),
+        axisAnalysisPass(axisAnalysisPass) {}
+
+  // Try to deduplicate the resultVals based on the
+  // constancy properties of the result discovered by
+  // the axis analysis pass. If possible, redundant
+  // computation is eliminated.
+  SmallVector<Value> maybeDeduplicate(SourceOp op,
+                                      SmallVector<Value> resultVals) const {
+    if (!isMemoryEffectFree(op))
+      // the op has side effects: can't dedup
+      return resultVals;
+    SmallVector<Value> results = op->getResults();
+    if (results.size() == 0 || results.size() > 1)
+      // there must be exactly 1 result
+      return resultVals;
+    Value result = results[0];
+    Type type = result.getType();
+    if (!type)
+      return resultVals;
+    RankedTensorType rtType = type.dyn_cast<RankedTensorType>();
+    if (!rtType)
+      // the result must be a tensor
+      return resultVals;
+    Attribute encoding = rtType.getEncoding();
+    if (!encoding)
+      // encoding not available
+      return resultVals;
+    triton::gpu::BlockedEncodingAttr blockedEncoding =
+        encoding.dyn_cast<triton::gpu::BlockedEncodingAttr>();
+    if (!blockedEncoding)
```

Comment:
Why doesn't it work for non-blocking encoding cases?

Reply:
With some trial and error, I've found out that if I don't constrain the encoding type to `BlcokedEncodingAttr` at this point, the downstream `triton::gpu::getElemsPerThread` crashes in the `test_fp8_dot_acc`. This is the error:

```
$ python3 -m pytest python/test/unit/language/test_core.py -k test_fp8_dot_acc
========================================================== test session starts ==========================================================
platform linux -- Python 3.11.5, pytest-7.4.2, pluggy-1.3.0
rootdir: /data/users/aakhundov/triton/python
collected 7000 items / 6992 deselected / 8 selected                                                                                     

python/test/unit/language/test_core.py Fatal Python error: Aborted

Current thread 0x00007f3f3defc480 (most recent call first):
  File "/data/users/aakhundov/triton/python/triton/compiler/compiler.py", line 167 in ttgir_to_llir
  File "/data/users/aakhundov/triton/python/triton/compiler/compiler.py", line 416 in <lambda>
  File "/data/users/aakhundov/triton/python/triton/compiler/compiler.py", line 517 in compile
  File "/data/users/aakhundov/triton/python/triton/runtime/jit.py", line 396 in launcher_body
  File "<string>", line 4 in matmul_kernel
  File "/data/users/aakhundov/triton/python/test/unit/language/test_core.py", line 3919 in test_fp8_dot_acc
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/python.py", line 194 in pytest_pyfunc_call
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_callers.py", line 77 in _multicall
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_hooks.py", line 493 in __call__
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/python.py", line 1792 in runtest
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/runner.py", line 169 in pytest_runtest_call
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_callers.py", line 77 in _multicall
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_hooks.py", line 493 in __call__
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/runner.py", line 262 in <lambda>
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/runner.py", line 341 in from_call
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/runner.py", line 261 in call_runtest_hook
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/runner.py", line 222 in call_and_report
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/runner.py", line 133 in runtestprotocol
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/runner.py", line 114 in pytest_runtest_protocol
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_callers.py", line 77 in _multicall
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_hooks.py", line 493 in __call__
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/main.py", line 350 in pytest_runtestloop
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_callers.py", line 77 in _multicall
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_hooks.py", line 493 in __call__
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/main.py", line 325 in _main
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/main.py", line 271 in wrap_session
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/main.py", line 318 in pytest_cmdline_main
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_callers.py", line 77 in _multicall
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pluggy/_hooks.py", line 493 in __call__
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/config/__init__.py", line 169 in main
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/_pytest/config/__init__.py", line 192 in console_main
  File "/home/aakhundov/.conda/envs/triton/lib/python3.11/site-packages/pytest/__main__.py", line 5 in <module>
  File "<frozen runpy>", line 88 in _run_code
  File "<frozen runpy>", line 198 in _run_module_as_main

Extension modules: numpy.core._multiarray_umath, numpy.core._multiarray_tests, numpy.linalg._umath_linalg, numpy.fft._pocketfft_internal, numpy.random._common, numpy.random.bit_generator, numpy.random._bounded_integers, numpy.random._mt19937, numpy.random.mtrand, numpy.random._philox, numpy.random._pcg64, numpy.random._sfc64, numpy.random._generator, torch._C, torch._C._fft, torch._C._linalg, torch._C._nested, torch._C._nn, torch._C._sparse, torch._C._special (total: 20)
Aborted (core dumped)
```

Also, a few lines below, I'm calling `blockedEncoding.getSizePerThread()` to fetch the size per thread. In principle, I could use `triton::gpu::getSizePerThread` instead, but just checked and can confirm that this leads to the same crash above in the `test_fp8_dot_acc`. If you have any tips on how to resolve the crash, I'm happy to lift the current constraint on the encoding type. Thanks!

Reply:
Probably because getSizePerThread returns wrong results

Reply:
Can you please leave a note in the code so we can review it later?

Reply:
Also, maybe it works fine already with slicedencoding?

Reply:
@Jokeren I've extended the coverage to `SliceEncodingAttr`, too. Also added a `TODO` comment for addressing the `test_fp8_dot_acc` crashing in future.